File: services/api/app/schemas/report_documents.py
"""Contracts for immutable REPORT PDF documents and public verification."""

from datetime import datetime
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class StrictSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ReportDocumentType(StrEnum):
    RECEIPT = "receipt"
    VERIFIED = "verified"


class ReportDocumentStatus(StrEnum):
    PENDING = "pending"
    READY = "ready"
    FAILED = "failed"
    REPLACED = "replaced"
    REVOKED = "revoked"


class ReportDocument(StrictSchema):
    id: UUID
    document_type: ReportDocumentType
    version: int
    status: ReportDocumentStatus
    delivery_status: str
    file_sha256: str | None = None
    issued_at: datetime | None = None
    created_at: datetime
    revocation_reason: str | None = None


class ReportDocumentList(StrictSchema):
    items: list[ReportDocument]


class ReportDocumentRevision(StrictSchema):
    document_type: ReportDocumentType
    reason: str = Field(min_length=3, max_length=500)


class ReportDocumentReason(StrictSchema):
    reason: str = Field(min_length=3, max_length=500)


class CitizenDocumentDeliveryRequest(StrictSchema):
    sender_phone_number: str = Field(min_length=8, max_length=32)
    ticket_number: str = Field(pattern=r"^LP-[0-9]{4}-[0-9]{4,}$")
    document_type: ReportDocumentType = ReportDocumentType.RECEIPT


class ReportDocumentVerification(StrictSchema):
    valid: bool
    ticket_number: str
    document_type: ReportDocumentType
    version: int
    village_name: str
    issued_at: datetime | None
    status: str
    file_sha256: str | None

File: apps/dashboard/lib/report-documents.ts
import { apiFetch } from "./admin.ts";

export type ReportDocument = {
  id: string;
  document_type: "receipt" | "verified";
  version: number;
  status: "pending" | "ready" | "failed" | "replaced" | "revoked";
  delivery_status: "pending" | "sent" | "failed" | "unknown";
  file_sha256?: string | null;
  issued_at?: string | null;
  revocation_reason?: string | null;
  created_at: string;
};

export const getReportDocuments = (reportId: string) =>
  apiFetch<{ items: ReportDocument[] }>(`/api/v1/reports/${reportId}/documents`);

export const createReceiptDocument = (reportId: string) =>
  apiFetch<ReportDocument>(`/api/v1/reports/${reportId}/documents/receipt`, { method: "POST" });

export const retryReportDocument = (reportId: string, documentId: string) =>
  apiFetch<ReportDocument>(`/api/v1/reports/${reportId}/documents/${documentId}/retry`, { method: "POST" });

export const reviseReportDocument = (reportId: string, documentType: ReportDocument["document_type"], reason: string) =>
  apiFetch<ReportDocument>(`/api/v1/reports/${reportId}/documents/revisions`, {
    method: "POST", body: JSON.stringify({ document_type: documentType, reason }),
  });

export const revokeReportDocument = (reportId: string, documentId: string, reason: string) =>
  apiFetch<ReportDocument>(`/api/v1/reports/${reportId}/documents/${documentId}/revoke`, {
    method: "POST", body: JSON.stringify({ reason }),
  });
